src/server/serve.rs

```diff
@@ -57,75 +59,62 @@ where
                 continue;
             }
         };
+
         // Pass the returned stream to the `TokioIo` wrapper to convert
         // the stream into a tokio-compatible I/O stream.
         let io = TokioIo::new(stream);
 
-        // Spawn a tokio task to serve multiple connections concurrently and push
-        // the returned `JoinHandle` into the handles vector.
-        handles.push({
-            // Clone the `router` so it can be moved into the task.
+        // Create a hyper service to serve the incoming connection.
+        let service = {
             let router = Arc::clone(&router);
-            // Clone the `state` so it can be moved into the task.
             let state = Arc::clone(&state);
 
-            task::spawn(serve_connection(
-                router,
-                state,
-                permit,
-                io,
-                response_timeout,
-            ))
-        });
+            service_fn(move |request| {
```

Comment:
I plan on writing a struct that impl's Service along with a custom FutureResponse type. I have a feeling it might clean up the code a bit and could even improve performance if the service can be passed to `serve_connection` as `&service`.

Reply:
I might keep the `service_fn` depending on how I feel about the implementation.